src/java/org/apache/cassandra/streaming/compress/CompressedInputStream.java

```diff
@@ -114,88 +153,74 @@ private void decompressNextChunk() throws IOException
-    {
-        if (current >= bufferOffset + buffer.length || validBufferBytes == -1)
-            decompressNextChunk();
-
-        assert current >= bufferOffset && current < bufferOffset + validBufferBytes;
-
-        return ((int) buffer[(int) (current++ - bufferOffset)]) & 0xff;
-    }
-
-    @Override
-    public int read(byte[] b, int off, int len) throws IOException
+    private void decompress(ByteBuffer compressed) throws IOException
     {
-        long nextCurrent = current + len;
-
-        if (current >= bufferOffset + buffer.length || validBufferBytes == -1)
-            decompressNextChunk();
-
-        assert nextCurrent >= bufferOffset;
-
-        int read = 0;
-        while (read < len)
+        final int compressedChunkLength = info.parameters.chunkLength();
+        int length = compressed.remaining();
+        // uncompress, if the buffer size is less than chunk size
+        final boolean releaseCompressedBuffer;
+        if (length - CHECKSUM_LENGTH < compressedChunkLength)
         {
-            int nextLen = Math.min((len - read), (int)((bufferOffset + validBufferBytes) - current));
-
-            System.arraycopy(buffer, (int)(current - bufferOffset), b, off + read, nextLen);
-            read += nextLen;
-
-            current += nextLen;
-            if (read != len)
-                decompressNextChunk();
+            buffer.clear();
+            compressed.limit(length - CHECKSUM_LENGTH);
+            info.parameters.getSstableCompressor().uncompress(compressed, buffer);
+            buffer.flip();
+            releaseCompressedBuffer = true;
         }
-
-        return len;
-    }
-
-    public void position(long position)
-    {
-        assert position >= current : "stream can only read forward.";
-        current = position;
-    }
-
-    private void decompress(byte[] compressed) throws IOException
-    {
-        // uncompress
-        if (compressed.length - checksumBytes.length < info.parameters.maxCompressedLength())
-            validBufferBytes = info.parameters.getSstableCompressor().uncompress(compressed, 0, compressed.length - checksumBytes.length, buffer, 0);
         else
         {
-            validBufferBytes = compressed.length - checksumBytes.length;
-            System.arraycopy(compressed, 0, buffer, 0, validBufferBytes);
+            BufferPool.put(buffer);
+            buffer = compressed;
+            buffer.limit(length - CHECKSUM_LENGTH);
+            releaseCompressedBuffer = false;
         }
-        totalCompressedBytesRead += compressed.length;
+        totalCompressedBytesRead += length;
 
         // validate crc randomly
         double crcCheckChance = this.crcCheckChanceSupplier.get();
         if (crcCheckChance > 0d && crcCheckChance > ThreadLocalRandom.current().nextDouble())
         {
-            int checksum = (int) checksumType.of(compressed, 0, compressed.length - checksumBytes.length);
+            ByteBuffer crcBuf = compressed.duplicate();
```

Comment:
For the current ticket, I"m going to leave the behavior as it is. If we want to change the behavior, we should open up a new ticket. In it's current state it's configurable by the operator, which is `ColumnFamilyStore#crcCheckChance`